kubectl-openebs/cli/command/get/volume_list.go

```diff
@@ -45,64 +47,107 @@ func NewCmdGetVolume() *cobra.Command {
 		Long:    volumesListCommandHelpText,
 		Run: func(cmd *cobra.Command, args []string) {
 			openebsNs, _ := cmd.Flags().GetString("openebs-namespace")
-			util.CheckErr(RunVolumesList(cmd, openebsNs, args), util.Fatal)
+			casType, _ := cmd.Flags().GetString("cas-type")
+			casType = strings.ToLower(casType)
+			util.CheckErr(RunVolumesList(openebsNs, casType, args), util.Fatal)
 		},
 	}
 	return cmd
 }
 
 // RunVolumesList lists the volumes
-func RunVolumesList(cmd *cobra.Command, openebsNs string, vols []string) error {
-	k8sClient, err := client.NewK8sClient("")
-	util.CheckErr(err, util.Fatal)
-	if openebsNs == "" {
-		nsFromCli, err := k8sClient.GetOpenEBSNamespace(util.CstorCasType)
-		if err != nil {
-			//return errors.Wrap(err, "Error determining the openebs namespace, please specify using \"--openebs-namespace\" flag")
-			return errors.New("no cstor volumes found in the cluster")
+func RunVolumesList(openebsNs, casType string, vols []string) error {
+	k8sClient, err := client.NewK8sClient(openebsNs)
+	if err != nil {
+		return err
+	}
+	var nsMap map[string]string
+	// 0. Figure out the openebsNs & casType mess!!
+	if openebsNs == "" && casType != "" {
+		openebsNs, err := k8sClient.GetOpenEBSNamespace(casType)
+		if err == nil {
+			// TODO: Verbose log for this estimated namespace
+			k8sClient.Ns = openebsNs
+		} else {
+			_, _ = fmt.Fprintf(os.Stderr, "couldn't figure out openebs-namespace for cas-type=%s\n", casType)
+			return err
+		}
+	} else if openebsNs == "" && casType == "" {
+		// show all volumes
+		nsMap, _ = k8sClient.GetOpenEBSNamespaceMap()
+	} else if openebsNs != "" && casType == "" {
+		// only show stuff which have CRs in openebsNs, if it matches with a cas-type
+		nsMap, _ = k8sClient.GetOpenEBSNamespaceMap()
+		// figure out the cas-type from openebsNs
+		for k, v := range nsMap {
+			if v == openebsNs {
+				casType = k
```

Comment:
- How are we determing which casType to show if both the storage engines are in same namespace?

Reply:
The above code would likely return one casType if both are in same namespace.

Reply:
Aha, I guess it's best to drop this conditional then, i.e. I should not show the volume if the `JivaVolume` or the `CStorVolumeConfig` doesn't exist.

Reply:
Also, we might need some logic which will show `"No cstor and jiva volumes found"` if both the storage engines are not installed.

Reply:
This is easy, if the size of the rows slice is zero at the end, I can just echo out the cas-type and figure out how to return an NZEC, maybe 1

Reply:
Yep. That would do it.

Reply:
But, if we dont have both the storage engines installed, the `GetOpenEBSNamespaceMap` will return an error which will terminate the command, so we would not reach the end to check the row slice length.

Reply:
good point, I shouldn't just skip the error

Reply:
```
❯ kubectl openebs get volumes
NAMESPACE   NAME                                       STATUS    VERSION    CAPACITY   STORAGE CLASS         ATTACHED   ACCESS MODE     ATTACHED NODE
N/A         pvc-193844d7-3bef-45a3-8b7d-ed3991391b45   N/A       N/A        5Gi        cstor-csi             Released   ReadWriteOnce   N/A
openebs     pvc-478a8329-f02d-47e5-8288-0c28b582be25   RW        2.9.0      4Gi        openebs-jiva-csi-sc   Released   ReadWriteOnce   minikube-2
openebs     pvc-b84f60ae-3f26-4110-a85d-bce7ec00dacc   Healthy   2.0.0-ee   20Gi       common-storageclass   Released   ReadWriteOnce   minikube-1

❯ kubectl openebs get volumes --openebs-namespace=cstor
NAMESPACE   NAME                                       STATUS   VERSION   CAPACITY   STORAGE CLASS         ATTACHED   ACCESS MODE     ATTACHED NODE
cstor       pvc-193844d7-3bef-45a3-8b7d-ed3991391b45   Init     2.9.0     5Gi        cstor-csi             Released   ReadWriteOnce   N/A
N/A         pvc-478a8329-f02d-47e5-8288-0c28b582be25   N/A      N/A       4Gi        openebs-jiva-csi-sc   Released   ReadWriteOnce   N/A
N/A         pvc-b84f60ae-3f26-4110-a85d-bce7ec00dacc   N/A      N/A       20Gi       common-storageclass   Released   ReadWriteOnce   N/A

```

This seems to be breaking.

Reply:
```
❯ kubectl openebs get volumes --cas-type=jiva --openebs-namespace=cstor
NAMESPACE   NAME                                       STATUS   VERSION   CAPACITY   STORAGE CLASS         ATTACHED   ACCESS MODE     ATTACHED NODE
N/A         pvc-478a8329-f02d-47e5-8288-0c28b582be25   N/A      N/A       4Gi        openebs-jiva-csi-sc   Released   ReadWriteOnce   N/A

```
This shows jiva volume even if the passed namespace doesnot have jiva resources.

Reply:
The resources are getting populated by only one namespace, other namespace resources are filled as `N/A` even though they exist.

Reply:
```
❯ kubectl openebs get volumes
couldn't figure out the namespace for jiva & cstor

❯ kubectl openebs get volumes --cas-type=cstor --openebs-namespace=openebs
NAMESPACE   NAME                                       STATUS    VERSION    CAPACITY   STORAGE CLASS         ATTACHED   ACCESS MODE     ATTACHED NODE
N/A         pvc-193844d7-3bef-45a3-8b7d-ed3991391b45   N/A       N/A        5Gi        cstor-csi             Released   ReadWriteOnce   N/A
openebs     pvc-b84f60ae-3f26-4110-a85d-bce7ec00dacc   Healthy   2.0.0-ee   20Gi       common-storageclass   Released   ReadWriteOnce   minikube-1

❯ kubectl openebs get volumes --cas-type=cstor --openebs-namespace=cstor
NAMESPACE   NAME                                       STATUS   VERSION   CAPACITY   STORAGE CLASS         ATTACHED   ACCESS MODE     ATTACHED NODE
cstor       pvc-193844d7-3bef-45a3-8b7d-ed3991391b45   Init     2.9.0     5Gi        cstor-csi             Released   ReadWriteOnce   N/A
N/A         pvc-b84f60ae-3f26-4110-a85d-bce7ec00dacc   N/A      N/A       20Gi       common-storageclass   Released   ReadWriteOnce   N/A

❯ kubectl openebs get volumes --cas-type=jiva --openebs-namespace=cstor
NAMESPACE   NAME                                       STATUS   VERSION   CAPACITY   STORAGE CLASS         ATTACHED   ACCESS MODE     ATTACHED NODE
N/A         pvc-478a8329-f02d-47e5-8288-0c28b582be25   N/A      N/A       4Gi        openebs-jiva-csi-sc   Released   ReadWriteOnce   N/A

❯ kubectl openebs get volumes --cas-type=jiva --openebs-namespace=openebs
NAMESPACE   NAME                                       STATUS   VERSION   CAPACITY   STORAGE CLASS         ATTACHED   ACCESS MODE     ATTACHED NODE
openebs     pvc-478a8329-f02d-47e5-8288-0c28b582be25   RW       2.9.0     4Gi        openebs-jiva-csi-sc   Released   ReadWriteOnce   minikube-2

```
Few observations, If we dont have openebs and jiva controllers installed, which means `jiva` and `openebs` itself is not installed then:-
- If we donot pass any ns and castype in flags, it says could not determine ns for jiva and cstor(expected)
- If we pass ns and castype it shows the volumes, which it should not as in the above we showed errors.
- It shows the volumes even if the namespace provided is incorrect.

Reply:
This can be fixed in one simple way, I'll just `List` all the JV, CV & CVA put it in a map[name]-> resource and then I'll be able to always show filled values irrespective of the wrongly given namespace.

Reply:
I won't take >1hour to code & test all of it.

Reply:
Thank you.

Reply:
Fixed
```bash
harsh @ brahmastra ~/.../github.com/openebs/openebsctl (vol_list_jiva)
└─ $ ▶ ./openebsctl get vol
NAMESPACE   NAME                                       STATUS    VERSION    CAPACITY   STORAGE CLASS         ATTACHED   ACCESS MODE     ATTACHED NODE
jiva        pvc-478a8329-f02d-47e5-8288-0c28b582be25   RW        2.9.0      4Gi        openebs-jiva-csi-sc   Released   ReadWriteOnce   minikube
openebs     pvc-b84f60ae-3f26-4110-a85d-bce7ec00dacc   Healthy   2.0.0-ee   20Gi       common-storageclass   Released   ReadWriteOnce   ip-10-0-6-143.ec2.internal
harsh @ brahmastra ~/.../github.com/openebs/openebsctl (vol_list_jiva)
└─ $ ▶ ./openebsctl get vol --cas-type=jiva
NAMESPACE   NAME                                       STATUS   VERSION   CAPACITY   STORAGE CLASS         ATTACHED   ACCESS MODE     ATTACHED NODE
jiva        pvc-478a8329-f02d-47e5-8288-0c28b582be25   RW       2.9.0     4Gi        openebs-jiva-csi-sc   Released   ReadWriteOnce   minikube
harsh @ brahmastra ~/.../github.com/openebs/openebsctl (vol_list_jiva)
└─ $ ▶ ./openebsctl get vol --cas-type=cstor
NAMESPACE   NAME                                       STATUS    VERSION    CAPACITY   STORAGE CLASS         ATTACHED   ACCESS MODE     ATTACHED NODE
openebs     pvc-b84f60ae-3f26-4110-a85d-bce7ec00dacc   Healthy   2.0.0-ee   20Gi       common-storageclass   Released   ReadWriteOnce   ip-10-0-6-143.ec2.internal
```